include/stl2/view/drop.hpp

```diff
@@ -47,36 +49,62 @@ STL2_OPEN_NAMESPACE {
 
 			constexpr R base() const { return base_; }
 
-			constexpr auto begin() requires !SimpleView<R> { return begin_impl(*this); }
-			constexpr auto begin() const requires Range<R const>
+			constexpr auto begin()
+			requires !(SimpleView<R> && RandomAccessRange<R>)
+			{ return begin_impl(*this); }
+			constexpr auto begin() const requires Range<R const> && RandomAccessRange<R const>
 			{ return begin_impl(*this); }
 
-			constexpr auto end() requires !SimpleView<R> { return end_impl(*this); }
-			constexpr auto end() const requires Range<R const>
+			constexpr auto end()
+			requires !(SimpleView<R> && RandomAccessRange<R>)
+			{ return end_impl(*this); }
+			constexpr auto end() const requires Range<R const> && RandomAccessRange<R const>
 			{ return end_impl(*this); }
 
 			constexpr auto size() requires !SimpleView<R> && SizedRange<R> { return size_impl(*this); }
 			constexpr auto size() const requires SizedRange<R const> { return size_impl(*this); }
 		private:
 			R base_;
 			D count_;
-			mutable detail::non_propagating_cache<iterator_t<R>> begin_;
+
+			constexpr auto& cached_begin() noexcept
+			{
+				return static_cast<typename drop_view::non_propagating_cache&>(*this);
+			}
 
 			template <class X>
 			static constexpr auto begin_impl(X& x)
 			{
-				if (!x.begin_) {
+				[[maybe_unused]] auto checked_begin = [&x]{
+					D const dist = __stl2::distance(x.base_);
+					STL2_EXPECT(x.count_ >= 0);
+					return __stl2::next(__stl2::begin(x.base_), dist < x.count_ ? dist : x.count_);
+				};
+
+				[[maybe_unused]] auto unchecked_begin = [&x]{
+					return __stl2::next(__stl2::begin(x.base_), x.count_);
+				};
+
+				if constexpr (RandomAccessRange<__maybe_const<is_const_v<X>, R>>) {
 					if constexpr (SizedRange<__maybe_const<is_const_v<X>, R>>) {
```

Comment:
You could merge the `XXX_begin` lambdas by moving this `if constexpr` switch inside.

Reply:
Aside: I *hate* that GCC and MSVC emit "unused" warnings for things that are only referenced in a discarded branch of an `if constexpr`. I must have put two dozen workarounds in the rework of `<variant>` I was working on this week.

Reply:
I did away with `checked_begin` and `unchecked_begin` when you mentioned that there's actually a bounded `next` (I thought you were saying "here's a good reason to propose one" earlier in the week).

I feel kind of sheepish having only an assert inside an `if constexpr` branch.

```cpp
auto compute_begin = [&x]{
	if constexpr (SizedRange<__maybe_const<is_const_v<X>, R>>) {
		STL2_ASSERT(x.count_ >= 0);
	}
	return __stl2::next(__stl2::begin(x.base_), x.count_, __stl2::end(x.base_));
};
```

On the plus side, no `[[maybe_unused]]` attribute is necessary!

Reply:
Yeah, I like this better. You can push that `ASSERT` up (and make it an `EXPECT` since it's "free", and may help the optimizer in release builds) and then "unbranch" some of the control flow.

Reply:
> Yeah, I like this better.

Oops - except that simply the bounded `next` call is O(n) for `SizedRange`s whose iterator and sentinel don't model `SizedSentinel` - we should probably keep the `if constexpr (SizedRange)` switch inside `compute_begin`.

Reply:
> ... we should probably keep the `if constexpr (SizedRange)` switch inside `compute_begin`

FYI, this is the only blocker now. I think the PR is good otherwise.